### PR TITLE
osd: partial recovery strategy based on PGLog

### PR DESCRIPTION
**Pull request by mslovy**
continue on https://github.com/ceph/ceph/pull/19569,  and switch to this PR

Signed-off-by: Yao Ning <yaoning@unitedstack.com>

### CODE REVIEW

**Comment by mslovy** [comment]
@jdurgin   we will continue on this and our another teammate will help on this, so switch to a new PR.  And problems solved base on previous PR.

**Comment by jdurgin** [comment]
@mslovy great, glad to see the update! I scheduled another rados run of this rebased on master:

http://pulpito.ceph.com/joshd-2018-04-30_18:22:59-rados-wip-partial-recovery-20180430-distro-basic-smithi/

**Comment by mslovy** [comment]
thanks, so those kinds of fails may false positive, we can ignore, right? 
Failure Reason:
Command crashed: 'CEPH_CLIENT_ID=0 adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage ceph_test_rados --max-ops 400000 --objects 1024 --max-in-flight 64 --size 4000000 --min-stride-size 400000 --max-stride-size 800000 --max-seconds 600 --op snap_remove 50 --op snap_create 50 --op rollback 50 --op setattr 25 --op read 100 --op copy_from 50 --op write 50 --op write_excl 50 --op rmattr 25 --op delete 50 --pool unique_pool_0

**Comment by letterwuyu** [comment]
@jdurgin Hi, Thanks for you great work.
I have updated the code, mainly repair some bug about object digest inconsistent , hope you can review and run QA so that according to the log to improve the code.

**Comment by jdurgin** [comment]
@letterwuyu you'll need to rebase since the ::decode functions were changed to use const_iterator

**Comment by letterwuyu** [comment]
@jdurgin ok, i have updated code.

**Comment by mslovy** [comment]
hi, @jdurgin , we fix out many bugs in the newest one, and rebase all commit to the master branch, can you help us to rerun teuthology.

**Comment by jdurgin** [comment]
here's a new run:

http://pulpito.ceph.com/joshd-2018-05-29_17:04:42-rados-wip-partial-recovery-2018-05-29-distro-basic-smithi/

**Comment by letterwuyu** [comment]
hi @jdurgin  I've been analyzing these logs these days，I find that a lot of failures seem to have nothing to do with us. like:
Command crashed: 'CEPH_CLIENT_ID=0 adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage ceph_test_rados --set_chunk --low_tier_pool low_tier --max-ops 4000 --objects 300 --max-in-flight 16 --size 4000000 --min-stride-size 400000 --max-stride-size 800000 --max-seconds 0 --op read 100 --op write 50 --op write_excl 50 --op delete 10 --pool unique_pool_0' 

I went to see the OSD log all this failure：
onn(0x55acc4c57600 :6809 s=STATE_ACCEPTING_WAIT_CONNECT_MSG pgs=0 cs=0 l=0).read_until read failed
n
 _read_bdev_label failed to open /var/lib/ceph/osd/ceph-0/block: (2) No such file or direct
ory

**Comment by letterwuyu** [comment]
Hi @jdurgin, I fix some bug, can you help me to rerun teuthology.

**Comment by jdurgin** [comment]
http://pulpito.ceph.com/joshd-2018-06-14_11:22:58-rados-wip-partial-recovery-20180613-distro-basic-smithi/

**Comment by letterwuyu** [comment]
Hi @jdurgin, I fix one bug and rebase all commit to the master branch, can you help me to rerun teuthology.

**Comment by tchaikov** [comment]
@mslovy @letterwuyu will include this PR in my next rados qa run.

**Comment by yuriw** [comment]
[wip-yuri5-testing-2018-08-13-1555](https://trello.com/c/eJxLjQqr/371-wip-yuri5-testing-2018-08-13-1555)

**Comment by neha-ojha** [comment]
http://pulpito.ceph.com/yuriw-2018-08-13_19:36:30-rados-wip-yuri5-testing-2018-08-13-1555-distro-basic-smithi/

**Comment by letterwuyu** [comment]
Thank you all.

**Comment by mslovy** [comment]
@yuriw  can we get the corresponding  ceph debuginfo package for a certain teuthology job?

**Comment by letterwuyu** [comment]
Hi, we fix two bug , please you help me to rerun teuthology, thinks.

**Comment by letterwuyu** [comment]
Makecahe reported "virtual memory exhausted: Cannot allocate memory" e rror, I think it has nothing to do with the code.

**Comment by letterwuyu** [comment]
Hi guys，please help me rerun teuthology, thanks.

**Comment by liewegas** [comment]
Can you rebase please?

**Comment by letterwuyu** [comment]
ok

**Comment by mslovy** [comment]
ok

**Comment by jdurgin** [comment]
http://pulpito.ceph.com/joshd-2018-09-18_02:55:59-rados-wip-partial-recovery-20180917-distro-basic-smithi/

**Comment by liewegas** [comment]
retest this please

**Comment by mslovy** [comment]
fix cache_tier hitset issues. can we rerun the teuthology?

**Comment by yuriw** [comment]
[wip-yuri5-testing-2018-10-26-1345](https://trello.com/c/bILd2oGM/453-wip-yuri5-testing-2018-10-26-1345)

**Comment by neha-ojha** [comment]
@mslovy See latest test results here: http://pulpito.ceph.com/yuriw-2018-10-26_19:41:04-rados-wip-yuri5-testing-2018-10-26-1345-distro-basic-smithi/

also, needs rebasing now.

**Comment by letterwuyu** [comment]
Hi, fix some  consistency problem after recovery, please help me to rerun teuthology.

**Comment by yuriw** [comment]
[wip-yuri-testing-2018-11-19-2035](https://trello.com/c/rXrPBkmM/477-wip-yuri-testing-2018-11-19-2035)

**Comment by neha-ojha** [comment]
@letterwuyu The latest test run looks much cleaner than earlier ones, but still has some issues http://pulpito.ceph.com/yuriw-2018-11-19_22:29:46-rados-wip-yuri-testing-2018-11-19-2035-distro-basic-smithi/

**Comment by letterwuyu** [comment]
OK，thx

**Comment by letterwuyu** [comment]
Hi, fix bug on scrub repair, please help me to rerun teuthology.

**Comment by yuriw** [comment]
[wip-yuri5-testing-2018-12-20-1934](https://trello.com/c/Wb8gooTk/495-wip-yuri5-testing-2018-12-20-1934)

**Comment by neha-ojha** [comment]
@letterwuyu Here's the latest test run http://pulpito.ceph.com/yuriw-2018-12-20_21:44:00-rados-wip-yuri5-testing-2018-12-20-1934-distro-basic-smithi/, a few of these are known failures, which have been fixed/are getting fixed. 

See if you find any related issues, else it might be a good idea to rebase and re-run through teuthology.

**Comment by letterwuyu** [comment]
@neha-ojha  Hi,I have rebase, but make check(arm 64) has an error that is not related to my code. please help me rerun teuthology.

**Comment by yuriw** [comment]
[wip-yuri5-testing-2019-01-02-2023](https://trello.com/c/yG4o4zxZ/496-wip-yuri5-testing-2019-01-02-2023)

**Comment by letterwuyu** [comment]
Hi，I have rebase and make check pass

**Comment by mslovy** [comment]
@jdurgin  those issues look like not related to our commit,  can you take a look again and any further test cases need to be added?

**Comment by jdurgin** [comment]
I agree, it looks like none of those failures are related. That's great news! I think the next step is code review for this PR

File: src/osd/PGLog.h [hunk 1]
```diff
@@ -1410,11 +1410,12 @@ struct PGLog : DoutPrefixProvider {
 	for (list<pg_log_entry_t>::reverse_iterator i = log.log.rbegin();
 	     i != log.log.rend();
 	     ++i) {
-	  if (!debug_verify_stored_missing && i->version <= info.last_complete) break;
 	  if (i->soid > info.last_backfill)
 	    continue;
 	  if (i->is_error())
 	    continue;
+    if (!i->is_delete())
+        missing.merge(*i);
```
Comment (Jeegn-Chen) [inline src/osd/PGLog.h]: My understanding is that read_log_and_missing() is only invoked this way: OSD::load_pgs() --> PG::read_state() --> PGLog::read_log_and_missing().
So missing.merge() here is just invoked when an OSD boots and more complicated during PG peering has not been addressed yet, right?

File: src/osd/ReplicatedBackend.cc [hunk 2]
```diff
@@ -1971,6 +2072,9 @@ int ReplicatedBackend::build_push_op(const ObjectRecoveryInfo &recovery_info,
     new_progress.data_complete = true;
     if (stat)
       stat->num_objects_recovered++;
+  } else if (progress.first && progress.omap_complete) {
+    // If omap is not changed, we need recovery omap when recovery cannot be completed once
+    new_progress.omap_complete = false;
```
Comment (Jeegn-Chen) [inline src/osd/ReplicatedBackend.cc]: Why do we need this? What limitation in the code make this logic necessary?

File: src/osd/PGLog.h (hunk 1 above)
Comment (letterwuyu) [reply src/osd/PGLog.h]: Sorry for replying so late，Yes, you are right.

File: src/osd/ReplicatedBackend.cc (hunk 2 above)
Comment (letterwuyu) [reply src/osd/ReplicatedBackend.cc]: If the object needs to be recovery multiple times, a temp object will be created, so even if the omap has not changed, we need to fill the omap of the temp object in submit_push_data.
For the case where multiple recovery is required and omap has not change , we can also populate the temp object with the omap of the need recovery object.

**Comment by letterwuyu** [comment]
Has rebased to resolve the submit_push_data code conflict

**Comment by mslovy** [comment]
anyone help review it, thx

**Comment by liewegas** [comment]
Now that nautilus is out, we should rebase this, review, test, and get it merged quickly so that we have the entirety of the octopus cycle to discover any bugs.

**Comment by neha-ojha** [comment]
@mslovy Could you please rebase this PR on current master?

**Comment by xiexingguo** [comment]
Seems rebase has already been done. Github shows no conflicts :-)

I agree with Sage. It would be great to have this in Octopus.

**Comment by neha-ojha** [comment]
@xiexingguo I'd like to run this through teuthology again, want to make sure things are updated, since we ran this a while back.

**Comment by xiexingguo** [comment]
> I'd like to run this through teuthology again, want to make sure things are updated, since we ran this a while back.

Sure, Neha:-)

**Comment by letterwuyu** [comment]
@neha-ojha Hi,I have changed commit info and rebase master

**Comment by letterwuyu** [comment]
rebase master

**Comment by letterwuyu** [comment]
please rerun teuthology

**Comment by neha-ojha** [comment]
@letterwuyu http://pulpito.ceph.com/nojha-2019-04-11_19:53:24-rados-wip-parial-recovery-2019-04-11-distro-basic-smithi/, none of the failures look related so far, there is quite a lot of noise in the run

**Comment by neha-ojha** [comment]
The failure in http://qa-proxy.ceph.com/teuthology/nojha-2019-04-11_19:53:24-rados-wip-parial-recovery-2019-04-11-distro-basic-smithi/3834491/teuthology.log looks related, unfortunately there are no logs because the job died.
```
2019-04-11T20:41:42.444 INFO:tasks.ceph.osd.3.smithi136.stderr:ceph-osd: /build/ceph-15.0.0-119-gb4f44ef/src/osd/ReplicatedBackend.cc:1609: void ReplicatedBackend::submit_push_data(const ObjectRecoveryInfo&, bool, bool, bool, bool, interval_set<long unsigned int>&, const interval_set<long unsigned int>&, ceph::bufferlist, ceph::bufferlist, const std::map<std::__cxx11::basic_string<char>, ceph::buffer::v14_2_0::list>&, const std::map<std::__cxx11::basic_string<char>, ceph::buffer::v14_2_0::list>&, ObjectStore::Transaction*): Assertion `r == 0' failed.
2019-04-11T20:41:42.444 INFO:tasks.ceph.osd.3.smithi136.stderr:*** Caught signal (Aborted) **
2019-04-11T20:41:42.444 INFO:tasks.ceph.osd.3.smithi136.stderr: in thread 7fa5c0d39700 thread_name:tp_osd_tp
2019-04-11T20:41:42.447 INFO:tasks.ceph.osd.3.smithi136.stderr: ceph version 15.0.0-119-gb4f44ef (b4f44ef7f8d9a3c5cef30600929599e43b975924) octopus (dev)
2019-04-11T20:41:42.447 INFO:tasks.ceph.osd.3.smithi136.stderr: 1: (()+0x12890) [0x7fa5e4677890]
2019-04-11T20:41:42.447 INFO:tasks.ceph.osd.3.smithi136.stderr: 2: (gsignal()+0xc7) [0x7fa5e3324e97]
2019-04-11T20:41:42.447 INFO:tasks.ceph.osd.3.smithi136.stderr: 3: (abort()+0x141) [0x7fa5e3326801]
2019-04-11T20:41:42.447 INFO:tasks.ceph.osd.3.smithi136.stderr: 4: (()+0x3039a) [0x7fa5e331639a]
2019-04-11T20:41:42.448 INFO:tasks.ceph.osd.3.smithi136.stderr: 5: (()+0x30412) [0x7fa5e3316412]
2019-04-11T20:41:42.448 INFO:tasks.ceph.osd.3.smithi136.stderr: 6: (ReplicatedBackend::submit_push_data(ObjectRecoveryInfo const&, bool, bool, bool, bool, interval_set<unsigned long, std::map<unsigned long, unsigned long, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, unsigned long> > > >&, interval_set<unsigned long, std::map<unsigned long, unsigned long, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, unsigned long> > > > const&, ceph::buffer::v14_2_0::list, ceph::buffer::v14_2_0::list, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, ceph::buffer::v14_2_0::list, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v14_2_0::list> > > const&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, ceph::buffer::v14_2_0::list, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, ceph::buffer::v14_2_0::list> > > const&, ObjectStore::Transaction*)+0x17da) [0x556c2de0f64a]
2019-04-11T20:41:42.448 INFO:tasks.ceph.osd.3.smithi136.stderr: 7: (ReplicatedBackend::handle_push(pg_shard_t, PushOp const&, PushReplyOp*, ObjectStore::Transaction*, bool)+0x2f5) [0x556c2de0fa95]
2019-04-11T20:41:42.448 INFO:tasks.ceph.osd.3.smithi136.stderr: 8: (ReplicatedBackend::_do_push(boost::intrusive_ptr<OpRequest>)+0x202) [0x556c2de0fed2]
2019-04-11T20:41:42.448 INFO:tasks.ceph.osd.3.smithi136.stderr: 9: (ReplicatedBackend::_handle_message(boost::intrusive_ptr<OpRequest>)+0x2a8) [0x556c2de17108]
2019-04-11T20:41:42.448 INFO:tasks.ceph.osd.3.smithi136.stderr: 10: (PGBackend::handle_message(boost::intrusive_ptr<OpRequest>)+0x57) [0x556c2dd251e7]
2019-04-11T20:41:42.448 INFO:tasks.ceph.osd.3.smithi136.stderr: 11: (PrimaryLogPG::do_request(boost::intrusive_ptr<OpRequest>&, ThreadPool::TPHandle&)+0x737) [0x556c2dcd8d87]
2019-04-11T20:41:42.449 INFO:tasks.ceph.osd.3.smithi136.stderr: 12: (OSD::dequeue_op(boost::intrusive_ptr<PG>, boost::intrusive_ptr<OpRequest>, ThreadPool::TPHandle&)+0x3dc) [0x556c2db0b95c]
2019-04-11T20:41:42.449 INFO:tasks.ceph.osd.3.smithi136.stderr: 13: (PGOpItem::run(OSD*, OSDShard*, boost::intrusive_ptr<PG>&, ThreadPool::TPHandle&)+0x62) [0x556c2dda6c92]
2019-04-11T20:41:42.449 INFO:tasks.ceph.osd.3.smithi136.stderr: 14: (OSD::ShardedOpWQ::_process(unsigned int, ceph::heartbeat_handle_d*)+0x7d9) [0x556c2db263a9]
2019-04-11T20:41:42.449 INFO:tasks.ceph.osd.3.smithi136.stderr: 15: (ShardedThreadPool::shardedthreadpool_worker(unsigned int)+0x5b4) [0x556c2e102a84]
2019-04-11T20:41:42.449 INFO:tasks.ceph.osd.3.smithi136.stderr: 16: (ShardedThreadPool::WorkThreadSharded::entry()+0x10) [0x556c2e104ae0]
2019-04-11T20:41:42.449 INFO:tasks.ceph.osd.3.smithi136.stderr: 17: (()+0x76db) [0x7fa5e466c6db]
2019-04-11T20:41:42.449 INFO:tasks.ceph.osd.3.smithi136.stderr: 18: (clone()+0x3f) [0x7fa5e340788f]
```
Rerunning failed/dead jobs: http://pulpito.ceph.com/nojha-2019-04-12_21:50:53-rados-wip-parial-recovery-2019-04-11-distro-basic-smithi/

**Comment by letterwuyu** [comment]
@neha-ojha Yes, it looks related

**Comment by letterwuyu** [comment]
@neha-ojha can you help me rerun, I need a log to confirm

**Comment by neha-ojha** [comment]
@letterwuyu That failure did not reproduce on rerun, sure, I can run this PR through another round of testing if that helps catch any latent bugs.

**Comment by letterwuyu** [comment]
@neha-ojha thx, Please run this PR through another round of testing,  It’s great if we get the log.

**Comment by neha-ojha** [comment]
Latest run: http://pulpito.ceph.com/nojha-2019-04-22_22:19:47-rados-wip-parial-recovery-2018-04-22-distro-basic-smithi/

@letterwuyu http://qa-proxy.ceph.com/teuthology/nojha-2019-04-22_22:19:47-rados-wip-parial-recovery-2018-04-22-distro-basic-smithi/3880459/teuthology.log looks same as https://github.com/ceph/ceph/pull/21722#issuecomment-482734559 - this time with logs

**Comment by letterwuyu** [comment]
@neha-ojha ok, great, thx very much

**Comment by letterwuyu** [comment]
@neha-ojha, hi, I fix this bug, please help me rerun test

**Comment by neha-ojha** [comment]
> @neha-ojha, hi, I fix this bug, please help me rerun test

yup

**Comment by letterwuyu** [comment]
have rebase mater，
the amd64 make check fail，look like Not enough memory
ceph_assert(mem_avail >= 0)
 ceph version Development (no_version) octopus (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x150) [0x2d1f418]

**Comment by neha-ojha** [comment]
@letterwuyu This is a rados run before your latest rebase http://pulpito.ceph.com/nojha-2019-04-25_21:33:30-rados-wip-parial-recovery-2019-04-25-distro-basic-smithi/

**Comment by letterwuyu** [comment]
@neha-ojha   it looks like none of those failures are related. Can we start review？

**Comment by letterwuyu** [comment]
“saw valgrind issues” looks like because a nullptr
https://github.com/ceph/ceph/blob/f4684eb864c4b2a573bd2f3bae5bb82cb63ae598/src/msg/async/crypto_onwire.cc#L267

**Comment by mslovy** [comment]
hi, @neha-ojha , already split the commits and pass the simple tests.  can anyone help to review it

**Comment by neha-ojha** [comment]
> hi, @neha-ojha , already split the commits and pass the simple tests. can anyone help to review it

great, we'll review it this week.

File: src/osd/osd_types.h [hunk 3]
```diff
@@ -4366,47 +4451,60 @@ class pg_missing_set : public pg_missing_const_i {
...
-      missing[oid].have = have;
+      (p->second).have = have;
     }
   }
 
   void add(const hobject_t& oid, eversion_t need, eversion_t have,
-	   bool is_delete) {
-    missing[oid] = item(need, have, is_delete);
+	   bool is_delete, bool make_dirty = true) {
+    missing[oid] = item(need, have, is_delete, make_dirty, have == eversion_t());
```
Comment (neha-ojha) [inline src/osd/osd_types.h]: using `have == eversion_t()` is a bit confusing here, maybe add a comment to explain it

File: src/osd/ReplicatedBackend.cc [hunk 4]
```diff
@@ -1570,11 +1615,46 @@ void ReplicatedBackend::submit_push_data(
...
     fadvise_flags |= CEPH_OSD_OP_FLAG_FADVISE_DONTNEED;
+  // Punch zeros for data, if fiemap indicates nothing but it is marked dirty
+  if (data_zeros.size() > 0) {
+    data_zeros.intersection_of(recovery_info.copy_subset);
+    assert(intervals_included.subset_of(data_zeros));
+    data_zeros.subtract(intervals_included);
+
+    dout(20) << __func__ <<" recovering object " << recovery_info.soid
+             << " copy_subset: " << recovery_info.copy_subset
+             << " intervals_inlcuded: " << intervals_included
```
Comment (neha-ojha) [inline src/osd/ReplicatedBackend.cc]: s/intervals_inlcuded/intervals_included/

File: src/osd/osd_types.h [hunk 5]
```diff
@@ -108,6 +108,8 @@
 /// priority when more full
 #define OSD_DELETE_PRIORITY_FULL 255
 
+/// max intervals for clean_offsets in ObjectCleanRegions
+#define OSD_CLEAN_OFFSETS_MAX_INTERVALS 10
```
[the remark this answers is not in the document]
Comment (letterwuyu) [reply src/osd/osd_types.h]: I agree with you, I will add a commit about it later.

File: src/osd/osd_types.cc [hunk 6]
```diff
@@ -4275,6 +4275,118 @@ void ObjectModDesc::decode(ceph::buffer::list::const_iterator &_bl)
...
+    }
+    clean_offsets.erase(shortest_interval);
+  }
+}
+
+void ObjectCleanRegions::merge(const ObjectCleanRegions &other)
+{
+  clean_offsets.intersection_of(other.clean_offsets);
+  clean_omap = clean_omap && other.clean_omap;
+  trim();
```
[the remark this answers is not in the document]
Comment (letterwuyu) [reply src/osd/osd_types.cc]: In this scenario, the priority_version of the pg log will be "0'0" about write, the clean_region will be mark_fully_dirty in add_next_event. Then write related pg log and delete related pg log merge is still full dirty.

File: doc/dev/osd_internals/partial_object_recovery.rst [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+In order to achieve the goals, two things should be done: 
+
+1. logging where the object is modified is necessary
+2. logging whether the object_map of an object is modified is also neccessary
+
+class ObjectCleanRegion is introduced to do what we want.
+clean_offsets is a variable of interval_set<uint64_t>
+and is used to indicate the unmodified content in an object.
+clean_omap is a variable of bool indicating whether object_map is modified.
+max_num_intervals is an upbound of the number of intervals in clean_offsets
```
[the remark this answers is not in the document]
Comment (letterwuyu) [reply doc/dev/osd_internals/partial_object_recovery.rst]: ok

File: src/osd/osd_types.h (hunk 3 above)
Comment (letterwuyu) [reply src/osd/osd_types.h]: ok

File: src/osd/ReplicatedBackend.cc (hunk 4 above)
Comment (letterwuyu) [reply src/osd/ReplicatedBackend.cc]: ok

File: src/osd/osd_types.h [hunk 8]
```diff
@@ -3744,6 +3746,51 @@ class ObjectModDesc {
...
+  ObjectCleanRegions(int32_t max = OSD_CLEAN_OFFSETS_MAX_INTERVALS) : new_object(false), max_num_intervals(max) {
+    clean_offsets.insert(0, (uint64_t)-1);
+    clean_omap = true;
+  }
+  ObjectCleanRegions(uint64_t offset, uint64_t len, bool co, int32_t max = OSD_CLEAN_OFFSETS_MAX_INTERVALS)
+    : clean_omap(co), new_object(false), max_num_intervals(max) {
+    clean_offsets.insert(offset, len);
+  }
+  bool operator==(const ObjectCleanRegions &orc) const {
+    return clean_offsets == orc.clean_offsets && clean_omap == orc.clean_omap && max_num_intervals == orc.max_num_intervals;
```
[the remark this answers is not in the document]
Comment (letterwuyu) [reply src/osd/osd_types.h]: ok

File: src/osd/osd_types.h [hunk 9]
```diff
@@ -4413,47 +4451,60 @@ class pg_missing_set : public pg_missing_const_i {
...
+	      rmissing.erase((missing_it->second).need.version);
+        // .have = nil
+	missing_it->second = item(e.version, eversion_t(), e.is_delete());
+	missing_it->second.clean_regions.mark_fully_dirty();
+      } else {
+         // create new element in missing map
+         // .have = nil
+        missing[e.soid] = item(e.version, eversion_t(), e.is_delete());
+        missing[e.soid].clean_regions = e.clean_regions;
+        missing[e.soid].clean_regions.mark_fully_dirty();
```
[the remark this answers is not in the document]
Comment (letterwuyu) [reply src/osd/osd_types.h]: ok

File: src/osd/osd_types.h [hunk 10]
```diff
@@ -4201,48 +4202,77 @@ inline std::ostream& operator<<(std::ostream& out, const pg_log_t& log)
...
-      // macros post-luminous.
-      eversion_t e;
-      encode(e, bl);
-      encode(need, bl);
-      encode(have, bl);
+    // encoding a zeroed eversion_t to differentiate between OSD_RECOVERY_DELETES、
+    // SERVER_NAUTILUS and legacy unversioned encoding - a need value of 0'0 is not
+    // possible. This can be replaced with the legacy encoding
+
+    bool have_recovery_deletes = HAVE_FEATURE(features, OSD_RECOVERY_DELETES);
```
[the remark this answers is not in the document]
Comment (letterwuyu) [reply src/osd/osd_types.h]: ok

File: src/osd/osd_types.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (jdurgin) [reply src/osd/osd_types.cc]: yes, that handles it nicely, thanks!

**Comment by jdurgin** [comment]
thanks for the quick fixes, they look right to me. looking through do_osd_ops now

**Comment by jdurgin** [comment]
ah 'make check' is failing the encoding tests due to removal of the pre-luminous encoding for the missing items. This wasn't persisted to disk until luminous, and at that point it is always re-encoded with the format introduced with RECOVERY_DELETES, so we just need to fix the encoding test.

Adding an empty file to the ceph-object-corpus submodule as archive/15.0.0-539-g191ab33faf/forward_incomapt/pg_missing_t should allow the test to pass. For reference, it's running https://github.com/ceph/ceph/blob/master/src/test/encoding/readable.sh#L48

**Comment by letterwuyu** [comment]
@neha-ojha I have maded some changes to the code based on your comments.

**Comment by letterwuyu** [comment]
@jdurgin OK，thx

**Comment by neha-ojha** [comment]
> @neha-ojha I have maded some changes to the code based on your comments.

Cool, thanks! I guess we just need to address the make check failure now. I am also planning to run this PR through some other test suites, just to make sure we catch any bugs before merging.

**Comment by letterwuyu** [comment]
> ah 'make check' is failing the encoding tests due to removal of the pre-luminous encoding for the missing items. This wasn't persisted to disk until luminous, and at that point it is always re-encoded with the format introduced with RECOVERY_DELETES, so we just need to fix the encoding test.
> 
> Adding an empty file to the ceph-object-corpus submodule as archive/15.0.0-539-g191ab33faf/forward_incomapt/pg_missing_t should allow the test to pass. For reference, it's running https://github.com/ceph/ceph/blob/master/src/test/encoding/readable.sh#L48

hi @jdurgin , Ask a question, I have created a commit for the ceph-object-corpus submodule according to your suggestion, but the "Unmodified Submodules" failed. What should be the correct steps?
I first need the commit related to the ceph-object-corpus submodule merge?

**Comment by jdurgin** [comment]
@letterwuyu it seems we may need to merge the submodule change to ceph-object-corpus before 'make check' picks it up here. can you open a PR for that in ceph-object-corpus?

**Comment by letterwuyu** [comment]
@jdurgin I created pr
https://github.com/ceph/ceph-object-corpus/pull/8
https://github.com/ceph/ceph/pull/28034

**Comment by jdurgin** [comment]
jenkins: retest this please

**Comment by jdurgin** [comment]
jenkins: retest this please

**Comment by letterwuyu** [comment]
ok, make check has been successful，make check (arm64) seems to have an unrelated failure.

**Comment by jdurgin** [comment]
great! let's see if it passes the other suites @neha-ojha mentioned

**Comment by neha-ojha** [comment]
Latest test runs of 5 suites: http://pulpito.ceph.com/?branch=wip-parial-recovery-2019-05-09
None of the failures in rados and upgrade look related.
Waiting on @batrick to review the others.

**Comment by batrick** [comment]
failures are unrelated

**Comment by liewegas** [comment]
Yay!

**Comment by letterwuyu** [comment]
Thanks everyone

**Comment by jianwei1216** [comment]
I recently encountered a problem during the functional verification of incremental data recovery based on ceph-v15.2.8 version. See if you have ever encountered:
   
    Problem: After the osd that is undergoing incremental recovery is restarted, the data recovery is rolled back from the incremental recovery to the entire object recovery.
 
    Steps to reproduce:
    1. ceph osd set noout
    2. vdbench 5000 iops 8K 7:3 randrw, run for 3 minutes
    3. stop osd.0 running for 30 minutes
    4. start osd.0  --> Note 1: At this time, it is a real incremental recovery
    5. During the incremental recovery, again to stop osd.0
    6. After stopping osd.0, immediately start osd.0 again --> Note 2: At this time, it becomes a whole object recovery

    I'm still looking at this part of the code, wanting to understand why the pg that was originally undergoing incremental recovery, after a short osd restart, turned into a whole object recovery.

**Comment by neha-ojha** [comment]
> I recently encountered a problem during the functional verification of incremental data recovery based on ceph-v15.2.8 version. See if you have ever encountered:
> 
> ```
> Problem: After the osd that is undergoing incremental recovery is restarted, the data recovery is rolled back from the incremental recovery to the entire object recovery.
> 
> Steps to reproduce:
> 1. ceph osd set noout
> 2. vdbench 5000 iops 8K 7:3 randrw, run for 3 minutes
> 3. stop osd.0 running for 30 minutes
> 4. start osd.0  --> Note 1: At this time, it is a real incremental recovery
> 5. During the incremental recovery, again to stop osd.0
> 6. After stopping osd.0, immediately start osd.0 again --> Note 2: At this time, it becomes a whole object recovery
> 
> I'm still looking at this part of the code, wanting to understand why the pg that was originally undergoing incremental recovery, after a short osd restart, turned into a whole object recovery.
> ```

Hi @jianwei1216, could you please create a tracker issue and add your findings in it?

**Comment by jianwei1216** [comment]
> > I recently encountered a problem during the functional verification of incremental data recovery based on ceph-v15.2.8 version. See if you have ever encountered:
> > ```
> > Problem: After the osd that is undergoing incremental recovery is restarted, the data recovery is rolled back from the incremental recovery to the entire object recovery.
> > 
> > Steps to reproduce:
> > 1. ceph osd set noout
> > 2. vdbench 5000 iops 8K 7:3 randrw, run for 3 minutes
> > 3. stop osd.0 running for 30 minutes
> > 4. start osd.0  --> Note 1: At this time, it is a real incremental recovery
> > 5. During the incremental recovery, again to stop osd.0
> > 6. After stopping osd.0, immediately start osd.0 again --> Note 2: At this time, it becomes a whole object recovery
> > 
> > I'm still looking at this part of the code, wanting to understand why the pg that was originally undergoing incremental recovery, after a short osd restart, turned into a whole object recovery.
> > ```
> 
> Hi @jianwei1216, could you please create a tracker issue and add your findings in it?

already open a new pull request : https://github.com/ceph/ceph/pull/21722

**Comment by neha-ojha** [comment]
> > > I recently encountered a problem during the functional verification of incremental data recovery based on ceph-v15.2.8 version. See if you have ever encountered:
> > > ```
> > > Problem: After the osd that is undergoing incremental recovery is restarted, the data recovery is rolled back from the incremental recovery to the entire object recovery.
> > > 
> > > Steps to reproduce:
> > > 1. ceph osd set noout
> > > 2. vdbench 5000 iops 8K 7:3 randrw, run for 3 minutes
> > > 3. stop osd.0 running for 30 minutes
> > > 4. start osd.0  --> Note 1: At this time, it is a real incremental recovery
> > > 5. During the incremental recovery, again to stop osd.0
> > > 6. After stopping osd.0, immediately start osd.0 again --> Note 2: At this time, it becomes a whole object recovery
> > > 
> > > I'm still looking at this part of the code, wanting to understand why the pg that was originally undergoing incremental recovery, after a short osd restart, turned into a whole object recovery.
> > > ```
> > 
> > 
> > Hi @jianwei1216, could you please create a tracker issue and add your findings in it?
> 
> already open a new pull request : #21722

you mean https://github.com/ceph/ceph/pull/42904, thanks!

**Pull request merged**